File: algorithms/medium/find_closest_node_to_given_two_nodes.cpp
class Solution {
public:
  int closestMeetingNode(vector<int> &edges, int node1, int node2) {
    auto mp1 = dfs(edges, node1, 0);
    auto mp2 = dfs(edges, node2, 0);
    int ans = -1, min_dis = INT_MAX;
    for (int i = 0; i < edges.size(); i++) {
      auto f1 = mp1.find(i), f2 = mp2.find(i);
      if (f1 != mp1.end() && f2 != mp2.end()) {
        int ma = max(f1->second, f2->second);
        if (min_dis > ma) {
          ans = i;
          min_dis = ma;
        }
      }
    }
    return ans;
  }
  unordered_map<int, int> dfs(vector<int> &edges, int curr, int dis) {
    unordered_map<int, int> mp;
    while (curr != -1 && mp.find(curr) == mp.end()) {
      mp.insert({curr, dis++});
      curr = edges[curr];
    }
    return mp;
  }
};

File: algorithms/medium/find_closest_node_to_given_two_nodes.ex
defmodule Solution do
  @spec closest_meeting_node(edges :: [integer], node1 :: integer, node2 :: integer) :: integer
  def closest_meeting_node(edges, node1, node2) do
    edges =
      edges
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {e, i}, acc -> Map.put(acc, i, e) end)

    mp1 = dfs(edges, node1, 0)
    mp2 = dfs(edges, node2, 0)

    0..(map_size(edges) - 1)
    |> Enum.reduce({-1, 100_000}, fn n, {ans, dis} ->
      ma = max(mp1[n], mp2[n])
      if ma < dis, do: {n, ma}, else: {ans, dis}
    end)
    |> elem(0)
  end

  def dfs(edges, curr, dis, mp \\ %{}) do
    mp = Map.put(mp, curr, dis)
    next = Map.get(edges, curr, -1)

    if next == -1 || mp[next] != nil do
      mp
    else
      dfs(edges, next, dis + 1, mp)
    end
  end
end
